### PR TITLE
unarchive: Fail when zipinfo is not available

### LINKED ISSUE #39029: unarchive: Failed when zipinfo is unavailable
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
unarchive

##### ANSIBLE VERSION
```
ansible 2.5.1
  config file = /Users/markchassy/bitbucket/orson_ansible/orson/ansible.cfg
  configured module search path = [u'/Users/markchassy/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/markchassy/bitbucket/virtual/venv27/lib/python2.7/site-packages/ansible
  executable location = /Users/markchassy/bitbucket/virtual/venv27/bin/ansible
  python version = 2.7.14 (default, Sep 25 2017, 09:53:22) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.37)]
```

##### CONFIGURATION
```
DEFAULT_ROLES_PATH(env: ANSIBLE_ROLES_PATH) = [u'/Users/markchassy/bitbucket/ansible_roles']
```

##### OS / ENVIRONMENT
Host: Mac HIghSierra
Guest: Alpine 3.7 (Virtual)

##### SUMMARY
Using the unarchive module while using ansible-playbook is failing on an alpine vm
Strangely it works with ansible-container on alpine containers
Also copying over the zip and using command works on the VM

##### STEPS TO REPRODUCE

```
- name: Unarchive global repo
  unarchive:
    src: "./files/repo.zip"
    dest: "/opt/{{ application_user }}"
    mode: 0755
```

##### EXPECTED RESULTS
The archive should get copied over an unzipped

##### ACTUAL RESULTS

```
TASK [alp_app_role : Unarchive global repo] ************************************************************************************************
task path: /Users/markchassy/bitbucket/ansible_roles/alp_app_role/tasks/main.yml:7
<192.168.56.102> ESTABLISH SSH CONNECTION FOR USER: alpine
<192.168.56.102> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=alpine -o ConnectTimeout=10 -o ControlPath=/Users/markchassy/.ansible/cp/1e87d34aea 192.168.56.102 '/bin/sh -c '"'"'echo ~ && sleep 0'"'"''
<192.168.56.102> (0, '/home/alpine\n', '')
<192.168.56.102> ESTABLISH SSH CONNECTION FOR USER: alpine
<192.168.56.102> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=alpine -o ConnectTimeout=10 -o ControlPath=/Users/markchassy/.ansible/cp/1e87d34aea 192.168.56.102 '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo /home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454 `" && echo ansible-tmp-1524153183.66-100263092970454="` echo /home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454 `" ) && sleep 0'"'"''
<192.168.56.102> (0, 'ansible-tmp-1524153183.66-100263092970454=/home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454\n', '')
Using module file /Users/markchassy/bitbucket/virtual/venv27/lib/python2.7/site-packages/ansible/modules/files/stat.py
<192.168.56.102> PUT /Users/markchassy/.ansible/tmp/ansible-local-53181dIbYIS/tmpDpNDAv TO /home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454/stat.py
<192.168.56.102> SSH: EXEC sftp -b - -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=alpine -o ConnectTimeout=10 -o ControlPath=/Users/markchassy/.ansible/cp/1e87d34aea '[192.168.56.102]'
<192.168.56.102> (0, 'sftp> put /Users/markchassy/.ansible/tmp/ansible-local-53181dIbYIS/tmpDpNDAv /home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454/stat.py\n', '')
<192.168.56.102> ESTABLISH SSH CONNECTION FOR USER: alpine
<192.168.56.102> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=alpine -o ConnectTimeout=10 -o ControlPath=/Users/markchassy/.ansible/cp/1e87d34aea 192.168.56.102 '/bin/sh -c '"'"'chmod u+x /home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454/ /home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454/stat.py && sleep 0'"'"''
<192.168.56.102> (0, '', '')
<192.168.56.102> ESTABLISH SSH CONNECTION FOR USER: alpine
<192.168.56.102> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=alpine -o ConnectTimeout=10 -o ControlPath=/Users/markchassy/.ansible/cp/1e87d34aea -tt 192.168.56.102 '/bin/sh -c '"'"'sudo -H -S  -p "[sudo via ansible, key=gmjpfcqzursfcjvcmmqcgppljdmmmfzl] password: " -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-gmjpfcqzursfcjvcmmqcgppljdmmmfzl; /usr/bin/python /home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454/stat.py'"'"'"'"'"'"'"'"' && sleep 0'"'"''
Escalation succeeded
<192.168.56.102> (0, '\r\n\r\n{"invocation": {"module_args": {"checksum_algorithm": "sha1", "get_checksum": true, "follow": true, "checksum_algo": "sha1", "path": "/opt/orson", "get_md5": null, "get_mime": true, "get_attributes": true}}, "stat": {"charset": "unknown", "uid": 0, "exists": true, "attr_flags": "", "woth": false, "isreg": false, "device_type": 0, "mtime": 1524150417.079507, "block_size": 4096, "inode": 2354690, "isgid": false, "size": 4096, "executable": true, "isuid": false, "readable": true, "version": null, "pw_name": "root", "gid": 0, "ischr": false, "wusr": true, "writeable": true, "mimetype": "unknown", "blocks": 8, "xoth": true, "islnk": false, "nlink": 3, "issock": false, "rgrp": true, "gr_name": "root", "path": "/opt/orson", "xusr": true, "atime": 1524150415.6895733, "isdir": true, "ctime": 1524150417.079507, "isblk": false, "wgrp": true, "xgrp": true, "dev": 2051, "roth": true, "isfifo": false, "mode": "0775", "rusr": true, "attributes": []}, "changed": false}\r\n', 'Shared connection to 192.168.56.102 closed.\r\n')
<192.168.56.102> PUT /Users/markchassy/bitbucket/ansible_roles/alp_app_role/files/repo.zip TO /home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454/source
<192.168.56.102> SSH: EXEC sftp -b - -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=alpine -o ConnectTimeout=10 -o ControlPath=/Users/markchassy/.ansible/cp/1e87d34aea '[192.168.56.102]'
<192.168.56.102> (0, 'sftp> put /Users/markchassy/bitbucket/ansible_roles/alp_app_role/files/repo.zip /home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454/source\n', '')
<192.168.56.102> ESTABLISH SSH CONNECTION FOR USER: alpine
<192.168.56.102> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=alpine -o ConnectTimeout=10 -o ControlPath=/Users/markchassy/.ansible/cp/1e87d34aea 192.168.56.102 '/bin/sh -c '"'"'chmod u+x /home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454/ /home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454/source && sleep 0'"'"''
<192.168.56.102> (0, '', '')
Using module file /Users/markchassy/bitbucket/virtual/venv27/lib/python2.7/site-packages/ansible/modules/files/unarchive.py
<192.168.56.102> PUT /Users/markchassy/.ansible/tmp/ansible-local-53181dIbYIS/tmpQX2NwR TO /home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454/unarchive.py
<192.168.56.102> SSH: EXEC sftp -b - -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=alpine -o ConnectTimeout=10 -o ControlPath=/Users/markchassy/.ansible/cp/1e87d34aea '[192.168.56.102]'
<192.168.56.102> (0, 'sftp> put /Users/markchassy/.ansible/tmp/ansible-local-53181dIbYIS/tmpQX2NwR /home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454/unarchive.py\n', '')
<192.168.56.102> ESTABLISH SSH CONNECTION FOR USER: alpine
<192.168.56.102> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=alpine -o ConnectTimeout=10 -o ControlPath=/Users/markchassy/.ansible/cp/1e87d34aea 192.168.56.102 '/bin/sh -c '"'"'chmod u+x /home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454/ /home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454/unarchive.py && sleep 0'"'"''
<192.168.56.102> (0, '', '')
<192.168.56.102> ESTABLISH SSH CONNECTION FOR USER: alpine
<192.168.56.102> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=alpine -o ConnectTimeout=10 -o ControlPath=/Users/markchassy/.ansible/cp/1e87d34aea -tt 192.168.56.102 '/bin/sh -c '"'"'sudo -H -S  -p "[sudo via ansible, key=tonpueoiztysgyrveffifcylohkdpzyn] password: " -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-tonpueoiztysgyrveffifcylohkdpzyn; /usr/bin/python /home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454/unarchive.py'"'"'"'"'"'"'"'"' && sleep 0'"'"''
Escalation succeeded
<192.168.56.102> (1, '\r\n\r\n{"exception": "  File \\"/tmp/ansible_m58DY_/ansible_modlib.zip/ansible/module_utils/basic.py\\", line 2798, in run_command\\n    cmd = subprocess.Popen(args, **kwargs)\\n  File \\"/usr/lib/python2.7/subprocess.py\\", line 390, in __init__\\n    errread, errwrite)\\n  File \\"/usr/lib/python2.7/subprocess.py\\", line 1025, in _execute_child\\n    raise child_exception\\n", "cmd": "-T -s /home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454/source", "failed": true, "rc": 2, "invocation": {"module_args": {"directory_mode": null, "force": null, "remote_src": false, "exclude": [], "owner": null, "follow": false, "group": null, "unsafe_writes": null, "keep_newer": false, "setype": null, "content": null, "serole": null, "extra_opts": [], "dest": "/opt/orson", "selevel": null, "original_basename": "repo.zip", "regexp": null, "src": "/home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454/source", "validate_certs": true, "list_files": false, "seuser": null, "creates": null, "delimiter": null, "mode": 493, "attributes": null, "backup": null}}, "msg": "[Errno 2] No such file or directory"}\r\n', 'Shared connection to 192.168.56.102 closed.\r\n')
<192.168.56.102> ESTABLISH SSH CONNECTION FOR USER: alpine
<192.168.56.102> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=alpine -o ConnectTimeout=10 -o ControlPath=/Users/markchassy/.ansible/cp/1e87d34aea 192.168.56.102 '/bin/sh -c '"'"'rm -f -r /home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454/ > /dev/null 2>&1 && sleep 0'"'"''
<192.168.56.102> (0, '', '')
The full traceback is:
  File "/tmp/ansible_m58DY_/ansible_modlib.zip/ansible/module_utils/basic.py", line 2798, in run_command
    cmd = subprocess.Popen(args, **kwargs)
  File "/usr/lib/python2.7/subprocess.py", line 390, in __init__
    errread, errwrite)
  File "/usr/lib/python2.7/subprocess.py", line 1025, in _execute_child
    raise child_exception

fatal: [192.168.56.102]: FAILED! => {
    "changed": false,
    "cmd": "-T -s /home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454/source",
    "invocation": {
        "module_args": {
            "attributes": null,
            "backup": null,
            "content": null,
            "creates": null,
            "delimiter": null,
            "dest": "/opt/orson",
            "directory_mode": null,
            "exclude": [],
            "extra_opts": [],
            "follow": false,
            "force": null,
            "group": null,
            "keep_newer": false,
            "list_files": false,
            "mode": 493,
            "original_basename": "repo.zip",
            "owner": null,
            "regexp": null,
            "remote_src": false,
            "selevel": null,
            "serole": null,
            "setype": null,
            "seuser": null,
            "src": "/home/alpine/.ansible/tmp/ansible-tmp-1524153183.66-100263092970454/source",
            "unsafe_writes": null,
            "validate_certs": true
        }
    },
    "msg": "[Errno 2] No such file or directory",
    "rc": 2
}
	to retry, use: --limit @/Users/markchassy/bitbucket/orson_ansible/orson/alpine_playbook.retry
```

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY

Fixes: #39029

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
changelogs/fragments/39029_unarchive.yml
lib/ansible/modules/unarchive.py

### CODE REVIEW

**Comment by Akasurde** [comment]
Before -

```
"msg": "[Errno 2] No such file or directory",
```

After - 

```
 "msg": "Failed to find required executable zipinfo in paths"
```

File: lib/ansible/modules/unarchive.py [hunk 1]
```diff
@@ -383,8 +383,12 @@ def files_in_archive(self):
         return self._files_in_archive
 
     def is_unarchived(self):
-        # BSD unzip doesn't support zipinfo listings with timestamp.
-        cmd = [self.zipinfocmd_path, '-T', '-s', self.src]
+        if self.zipinfocmd_path:
+            # BSD unzip doesn't support zipinfo listings with timestamp.
+            cmd = [self.zipinfocmd_path, '-T', '-s', self.src]
+        else:
+            cmd = [self.cmd_path, '-Z', self.src]
```
Comment (Akasurde) [inline lib/ansible/modules/unarchive.py]: @bcoca Could you please test if `unzip -Z <archive>` works on BSD? I don't have BSD machines handy.

Comment (samdoran) [reply lib/ansible/modules/unarchive.py]: Doesn't look like it.
```
root@freebsd:~$ unzip -Z test.zip
Zipinfo mode needs additional options
```

Comment (samdoran) [reply lib/ansible/modules/unarchive.py]: Only `Z1` mode is supported, so this works.

```
root@freebsd:~$ unzip -Z1 test.zip
1.txt
2.txt
```

Comment (Akasurde) [reply lib/ansible/modules/unarchive.py]: It won't be enough, we need something like - 
```
# unzip -Z jce_policy-6.zip
Archive:  jce_policy-6.zip
Zip file size: 9101 bytes, number of entries: 5
drwxr-xr-x  2.2 unx        0 bx stor 06-Nov-17 07:40 jce/
-r--r--r--  2.2 unx     2663 tx defN 06-Nov-17 07:40 jce/COPYRIGHT.html
-r--r--r--  2.2 unx     8386 tx defN 06-Nov-17 07:40 jce/README.txt
-rw-r--r--  2.2 unx     2465 bx defN 06-Nov-17 07:40 jce/US_export_policy.jar
-rw-r--r--  2.2 unx     2481 bx defN 06-Nov-17 07:40 jce/local_policy.jar
5 files, 15995 bytes uncompressed, 8369 bytes compressed:  47.7%
```

**Comment by samdoran** [comment]
I'm not seeing how this error message will be displayde based on the changes in this PR.

```
 "msg": "Failed to find required executable zipinfo in paths"
```

That error message comes from `get_bin_path()` but only if the caller specifies `required=True`. That doesn't seem to be the case.

https://github.com/ansible/ansible/blob/20ef733ee02ba688757998404c1926381356b031/lib/ansible/modules/unarchive.py#L293

This PR seems to make the module work in the absence of `zipinfo` rather than updating the error message.

The module docs seem to indicate `zipinfo` is required so we should probably change the calls from `self.module.get_bin_path()` to `get_bin_path()` directly in `ansible.module_utils.common.process`. This raises an `ValueError` when the file is missing. We could catch and handle those, and fail the module early.

**Comment by samdoran** [comment]
Or if we can make the module work _without_ `zipinfo`, do that but just update the docs.

File: lib/ansible/modules/unarchive.py (hunk 1 above)
Comment (samdoran) [reply lib/ansible/modules/unarchive.py]: Yeah, then we'll have to fail if `zipinfo` is missing then.

**Review by samdoran — COMMENTED** [review]
Changes look good. Would it be possible to add tests for the missing binaries?

**Comment by Akasurde** [comment]
@samdoran I added tests, please take a look. Thanks

File: test/units/modules/test_unarchive.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            "include": "",
+        }
+        with pytest.raises(FailJson) as exec_info:
+            z = ZipArchive(
+                src="",
+                b_dest="",
+                file_args="",
+                module=fake_ansible_module,
+            )
+        assert "Unable to find required" in exec_info.value.kwargs["msg"]
```
Comment (samdoran) [inline test/units/modules/test_unarchive.py:57]: Using the `match` parameter saves you from having to do an `assert`.
```suggestion
        with pytest.raises(FailJson, match='Unable to find required'):
            z = ZipArchive(
                src="",
                b_dest="",
                file_args="",
                module=fake_ansible_module,
            )
```

File: test/units/modules/test_unarchive.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        }
+        fake_ansible_module.check_mode = False
+        with pytest.raises(FailJson) as exec_info:
+            z = TgzArchive(
+                src="",
+                b_dest="",
+                file_args="",
+                module=fake_ansible_module,
+            )
+        assert "Unable to find required" in exec_info.value.kwargs["msg"]
```
Comment (samdoran) [inline test/units/modules/test_unarchive.py:76]:
```suggestion
        with pytest.raises(FailJson, match='Unable to find required'):
            z = TgzArchive(
                src="",
                b_dest="",
                file_args="",
                module=fake_ansible_module,
            )
```

**Review by samdoran — APPROVED** [review]
Tests look good. Just one minor suggestion but not a blocker.

File: test/units/modules/test_unarchive.py (hunk 2 above)
Comment (Akasurde) [reply test/units/modules/test_unarchive.py:57]: That's dope.

File: test/units/modules/test_unarchive.py (hunk 3 above)
Comment (Akasurde) [reply test/units/modules/test_unarchive.py:76]: TIL

File: test/units/modules/test_unarchive.py (hunk 2 above)
Comment (samdoran) [reply test/units/modules/test_unarchive.py:57]: 😄

**Comment by felixfontein** [comment]
@Akasurde @samdoran this PR makes `unarchive` fail every time `unzip` is not installed, since `pick_handler` instantiates **every** handler beginning with `ZipArchive` until it finds one that can handle the archive. As an example, the RHEL yarn integration tests in community.general now fail (https://dev.azure.com/ansible/community.general/_build/results?buildId=16750&view=logs&j=b943b615-bf9c-5f33-798e-56b08a646841&t=6e59ed35-7427-5829-74c4-3ac929d25e26) when processing a `.tar.gz` file.

**Pull request merged**